cmd/hil-vpnd/mockprivops_test.go

```diff
@@ -0,0 +1,169 @@
+}
+
+// Generate a mock "key".
+func genMockKey() (string, error) {
+	var key [128 / 8]byte
+	_, err := rand.Read(key[:])
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("%x", key), nil
+}
+
+// Get a vpnInfo for the named vpn, panicing if it doesn't exist.
+func (ops *MockPrivOps) mustGetVpn(name string) *vpnInfo {
+	vpn, ok := ops.vpns[name]
+	if !ok {
+		panic(fmt.Sprintf("Tried to operate on non-existent vpn %q", name))
+	}
+	return vpn
+}
+
+//// Implementations of the methods needed to implement the PrivOps interface.
+
+func (ops *MockPrivOps) CreateVPN(name string, vlanNo uint16, portNo uint16) (string, error) {
+	ops.startOp()
+	defer ops.endOp()
+	if _, ok := ops.vpns[name]; ok {
+		panic(fmt.Sprintf(
+			"Tried to create a vpn with the same name (%q) as an existing one.",
+			name,
+		))
+	}
+
+	key, err := genMockKey()
+	if err != nil {
+		return "", err
+	}
+
+	ops.vpns[name] = &vpnInfo{
+		portNo: portNo,
+		vlanNo: vlanNo,
+		key:    key,
+	}
+
+	return key, nil
+}
+
+func (ops *MockPrivOps) StartVPN(name string) error {
+	ops.startOp()
+	defer ops.endOp()
+	vpn := ops.mustGetVpn(name)
+	if vpn.running {
+		panic(fmt.Sprintf("Tried to start already-running vpn %q", name))
+	}
+	vpn.running = true
+	return nil
+}
+
+func (ops *MockPrivOps) StopVPN(name string) error {
+	ops.startOp()
+	defer ops.endOp()
+	vpn := ops.mustGetVpn(name)
+	if !vpn.running {
+		panic(fmt.Sprintf("Tried to stop vpn %q, which is not running.", name))
+	}
+	vpn.running = false
+	return nil
+}
+
+func (ops *MockPrivOps) DeleteVPN(name string) error {
+	ops.startOp()
+	defer ops.endOp()
+	if _, ok := ops.vpns[name]; !ok {
+		panic(fmt.Sprintf("Tried to delete non-existent vpn %q", name))
+	}
+	delete(ops.vpns, name)
+	return nil
+}
+
+func (ops *MockPrivOps) ListVPNs() ([]string, error) {
+	ops.startOp()
+	defer ops.endOp()
+	ret := make([]string, 0, len(ops.vpns))
+	for k, _ := range ops.vpns {
+		ret = append(ret, k)
+	}
+	return ret, nil
+}
+
+//// Internal consistency stuff.
+
+// Call this at the start of every privileged operation; it locks the ops
+// value and does a consistency check.
+func (ops *MockPrivOps) startOp() {
+	ops.lock.Lock()
+	ops.validate()
+}
+
+// Call this at the end of every privileged operation; it does a consistency
+// check and unlocks the ops value.
+func (ops *MockPrivOps) endOp() {
+	ops.validate()
+	ops.lock.Unlock()
+}
+
+// Sanity check the state; panics if any of our invariants are violated.
+func (ops *MockPrivOps) validate() {
+	// Keep track of what port numbers we've seen as we walk through
+	// the set of vpns.
+	usedPorts := make(map[uint16]struct{})
+
+	for k, v := range ops.vpns {
+		// Make sure the port number isn't used by any vpn we've seen in
+		// the past.
+		if _, ok := usedPorts[v.portNo]; ok {
+			panic(fmt.Sprintf(
+				"Port number %d is used by more than one vpn!",
+				v.portNo,
+			))
+		}
+		// OK, we're good. Add this to the list for later checks:
+		usedPorts[v.portNo] = struct{}{}
```

Comment:
For the real PrivOps, we will be tracking this in a database, correct?

Reply:
I wasn't totally decided on this; if we ignore it, the worst that happens is that starting one of the vpns will fail, which doesn't present a security risk. And I'm not in love with adding a database to the privop command, as I want to keep it as simple as possible. So the idea here was that just as part of our testing for vpnd, we would make sure it doesn't violate this invariant itself.

Reply:
My current thinking is that on startup we do a privops list to get all existent vpns, and when creating them we put the port number in the name, so we can infer it from the names given to us by privop when we start back up. At runtime we can just keep track of the set of in-use port numbers. I actually think we don't even need a db for the vpn stuff at all.

Probably we want to have a config file that specifies allowed port ranges.

Reply:
okay, sounds good to me then.